modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsDestination.java

```diff
@@ -137,45 +222,72 @@ private void setKerberosAuth(Configuration configuration) throws IOException {
         UserGroupInformation.loginUserFromKeytab(options.getKerberosPrincipal(), options.getKerberosKeytabFile());
     }
 
+
+    protected void updateFilesTimer(HdfsFile hdfsfile) {
+        if (options.getTimeReap() == 0)
+           return;
+
+        if (timeReap != null)
+           return;
+
+        timeReap = new Timer();
+        timeReap.schedule( new TimeReapTask(), options.getTimeReap());
+    }
+
     @Override
     public boolean send(LogMessage logMessage) {
         isOpened = false;
         String resolvedFileName = options.getFileNameTemplate().getResolvedString(logMessage);
-        FSDataOutputStream fsDataOutputStream = getFSDataOutputStream(resolvedFileName);
-        if (fsDataOutputStream == null) {
+        lock.lock();
+        HdfsFile hdfsfile = getHdfsFile(resolvedFileName);
+        if (hdfsfile == null) {
             // Unable to open file
-            closeAll(archiveDir != null);
+            closeAll(true);
             return false;
         }
 
         try {
             String formattedMessage = options.getTemplate().getResolvedString(logMessage);
+
             logger.debug("Outgoing message: " + formattedMessage);
-            fsDataOutputStream.write(formattedMessage.getBytes(Charset.forName("UTF-8")));
+
+            byte[] rawMessage = formattedMessage.getBytes(Charset.forName("UTF-8"));
+
+            hdfsfile.write(rawMessage);
+
+            updateFilesTimer(hdfsfile);
         } catch (IOException e) {
             printStackTrace(e);
             closeAll(false);
             return false;
+        } finally {
+          lock.unlock();
         }
 
+
         isOpened = true;
         return true;
     }
 
-    private FSDataOutputStream getFSDataOutputStream(String resolvedFileName) {
+    private HdfsFile getHdfsFile(String resolvedFileName) {
         HdfsFile hdfsFile = openedFiles.get(resolvedFileName);
         if (hdfsFile == null) {
```

Comment:
this could be done easier: hdfsFile == null || !hdfsFile.isOpen()

Reply:
Well for me the two separate condition was more readable, but @gaborznagy also commented this, so I guess it is just me and time to change this part.

Reply:
the main problem is the code duplication, within the two if, you are doing the same thing twice

Reply:
Sure I figured what is bugging you.
Of course it is a code duplication but very local - really hard to miss - thus I won't think it would have caused any issue later on. But as I said I am going to change it as it seems its just me who preferred it that way.